esmvalcore/preprocessor/_multimodel.py

```diff
@@ -248,115 +128,179 @@ def _unify_time_coordinates(cubes):
-        times = [cube.coord('time').points for cube in cubes]
-        overlap = reduce(np.intersect1d, times)
-        if len(overlap) <= 1:
-            logger.info("Time overlap between cubes is none or a single point."
-                        "check datasets: will not compute statistics.")
-            return cubes
-        logger.debug("Using common time overlap between "
-                     "datasets to compute statistics.")
+        common_time_points = reduce(np.intersect1d, all_time_arrays)
+        new_cubes = [_subset(cube, common_time_points) for cube in cubes]
     elif span == 'full':
-        logger.debug("Using full time spans to compute statistics.")
+        all_time_points = reduce(np.union1d, all_time_arrays)
+        new_cubes = [_extend(cube, all_time_points) for cube in cubes]
     else:
-        raise ValueError(
-            "Unexpected value for span {}, choose from 'overlap', 'full'".
-            format(span))
+        raise ValueError(f"Invalid argument for span: {span!r}"
+                         "Must be one of 'overlap', 'full'.")
+
+    for cube in new_cubes:
+        # Make sure bounds exist and are consistent
+        cube.coord('time').bounds = None
+        cube.coord('time').guess_bounds()
+
+    return new_cubes
+
+
+def _combine(cubes):
+    """Merge iris cubes into a single big cube with new dimension.
+
+    This assumes that all input cubes have the same shape.
+    """
+    equalise_attributes(cubes)  # in-place
+
+    for i, cube in enumerate(cubes):
+        concat_dim = iris.coords.AuxCoord(i, var_name=CONCAT_DIM)
+
+        cube.add_aux_coord(concat_dim)
+
+        # Clear some metadata that can cause merge to fail
+        # https://scitools-iris.readthedocs.io/en/stable/userguide/
+        #    merge_and_concat.html#common-issues-with-merge-and-concatenate
+
+        cube.cell_methods = None
+
+        for coord in cube.coords():
+            coord.long_name = None
+            coord.attributes = None
+
+    cubes = iris.cube.CubeList(cubes)
+
+    merged_cube = cubes.merge_cube()
+
+    return merged_cube
+
+
+def _compute_eager(cubes: list, *, operator: iris.analysis.Aggregator,
+                   **kwargs):
+    """Compute statistics one slice at a time."""
+    _ = [cube.data for cube in cubes]  # make sure the cubes' data are realized
+
+    result_slices = []
+    for i in range(cubes[0].shape[0]):
+        single_model_slices = [cube[i] for cube in cubes]
+        combined_slice = _combine(single_model_slices)
+        collapsed_slice = combined_slice.collapsed(CONCAT_DIM, operator,
+                                                   **kwargs)
+
+        # some iris aggregators modify dtype, see e.g.
+        # https://numpy.org/doc/stable/reference/generated/numpy.ma.average.html
+        collapsed_slice.data = collapsed_slice.data.astype(np.float32)
+
+        result_slices.append(collapsed_slice)
+
+    result_cube = iris.cube.CubeList(result_slices).merge_cube()
```

Comment:
can this fail now? or we've already processed those cubes to a point where it will never fail

Reply:
I would say it shouldn't, but apparently it can fail indeed. E.g. if you calculate the mean of integer arrays `[1, 2]` and `[3, 3]` one index at a time, you'll end up concatenating `int(2)` and `float(2.5)`, which fails. Normally all input should be dtype `float32` and if dtype is preserved in the calculations, that shoudn't happen. 

Just in case, I'll add a try/except clause here as well, with a nice message.